Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -447,17 +447,35 @@ void GCMemcardManager::DrawIcons()
   const auto column = 3;
   for (int slot = 0; slot < SLOT_COUNT; slot++)
   {
-    int row = 0;
-    for (auto& icon : m_slot_active_icons[slot])
+    // skip loop if the table is empty
+    if (m_slot_table[slot]->rowCount() <= 0)
+      continue;
+
+    const auto viewport = m_slot_table[slot]->viewport();
+    auto row = m_slot_table[slot]->indexAt(viewport->rect().topLeft()).row();
+    const auto max_table_index = m_slot_table[slot]->indexAt(viewport->rect().bottomLeft());
+    const auto max_row =
+        max_table_index.row() < 0 ? (m_slot_table[slot]->rowCount() - 1) : max_table_index.row();
+
+    for (; row <= max_row; row++)
     {
-      int frame = (m_current_frame / 3 - icon.first) % icon.second.size();
+      auto& icon = m_slot_active_icons[slot][row];
```

Comment:
Thoughts on using structured bindings to give the individual `std::pair` members a name? (Just like deconstruction works in C#)

```c++
auto& [frame_count, frames] = m_slot_active_icons[slot][row];

// this icon doesn't have an animation
if (frames.size() <= 1)
  continue;

const auto prev_frame = ((m_current_frame - 1) / 3 - frame_count) % frames.size();
// etc.
```

I personally find them easier readable than `x.first`/`x.second` (or `x.Item1`/`x.Item2` in C#); and AFAICT most other projects I know tend to agree.

Reply:
i fully agree, and their name change in the other PR handling the animations ( icon.m_frames , icon.m_frame_timing etc )